### PR TITLE
borgbackup: 1.2.2 -> 1.2.3

### PR DESCRIPTION
**Pull request by mweinelt**
https://github.com/borgbackup/borg/blob/1.2.3/docs/changes.rst#version-123-2022-12-24

Currently reverting two commits that make a test fail, still have to talk to the upstream about that.

```
_______________ HashIndexSizeTestCase.test_size_on_disk_accurate _______________

self = <borg.testsuite.hashindex.HashIndexSizeTestCase testMethod=test_size_on_disk_accurate>

    def test_size_on_disk_accurate(self):
        idx = ChunkIndex()
        for i in range(1234):
            idx[H(i)] = i, i**2, i**3
        with tempfile.NamedTemporaryFile() as file:
            idx.write(file)
            size = os.path.getsize(file.fileno())
>       assert idx.size() == size
E       assert 90350 == 0
E        +  where 90350 = <built-in method size of borg.hashindex.ChunkIndex object at 0x7ffff351da10>()
E        +    where <built-in method size of borg.hashindex.ChunkIndex object at 0x7ffff351da10> = <borg.hashindex.ChunkIndex object at 0x7ffff351da10>.size
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mattmelling** [comment]
Result of `nixpkgs-review pr 207631` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>borgbackup</li>
    <li>borgmatic</li>
    <li>pika-backup</li>
    <li>vorta</li>
  </ul>
</details>

**Review by mattmelling — APPROVED** [review]

File: pkgs/tools/backup/borgbackup/default.nix [hunk 1]
```diff
@@ -7,21 +7,37 @@
...
+    (fetchpatch {
+      url = "https://github.com/borgbackup/borg/commit/05997e2867b2abb826abf9e9a564e288302bce8c.patch";
+      hash = "sha256-NORv1XAmM39hXOz4AxbByzydZ5Zn3NbiWJeZHmLb0hI=";
+      revert = true;
+    })
+    (fetchpatch {
+      url = "https://github.com/borgbackup/borg/commit/cc0ad321dc32b78ce2f2f625ae91040fddf3fd8c.patch";
+      hash = "sha256-kM9TfjMJkmGr3XZGB85HpcwuBJV8Y/vHEGnlDMUhKso=";
+      revert = true;
+    })
```
Comment (puetzk) [inline pkgs/tools/backup/borgbackup/default.nix:38]: i.e. to revert this change using the commit from the (1.2-maint) branch rather than the similar one from 2.0.0b4

```suggestion
    (fetchpatch {
      url = "https://github.com/borgbackup/borg/commit/8ddfd94f0904449175151faed67f9640710b6454.patch";
      hash = "sha256-+6ELPrvCTb21my6WanpqbPl2Sl1oxhb/z1zpo1KbMhc=";
      revert = true;
    })
```

Comment (puetzk) [reply pkgs/tools/backup/borgbackup/default.nix:38]: but I don't see any pytest failure to begin with, even with that change left in (i.e. no patching)

Comment (mweinelt) [reply pkgs/tools/backup/borgbackup/default.nix:38]: The failure is a race condition. The actual fix would be to to call `flush()` and `fsync()`.

Comment (puetzk) [reply pkgs/tools/backup/borgbackup/default.nix:38]: Ah, and so the key difference is in 
https://github.com/borgbackup/borg/blob/c29d4a096bba4018e14208b54fe24bdc886f4f3d/src/borg/hashindex.pyx#L123-L128

- given a str, it's used as a pathname to `with open(path, 'wb') as fd:`, and the file is closed before ChunkIndex.write returns. Given a file object, the write goes to the existing file object, which may not have flushed to the filesystem yet.
I don't think `fsync` should relevant - this doesn't care if the data is safe at rest on the physical medium. But it definitely matters whether or not cpython has actually made the relevant `write` syscall yet, or whether it's just spooling in-process.

Comment (puetzk) [reply pkgs/tools/backup/borgbackup/default.nix:38]: Ah, it will prefer `st_blksize` over `DEFAULT_BUFFER_SIZE` if the filesystem where the tempfile lives answers this stat: https://github.com/python/cpython/blob/024ac542d738f56b36bdeb3517a10e93da5acab9/Lib/_pyio.py#L247-L255 or (on most platforms) similar C code: https://github.com/python/cpython/blob/85dd6cb6df996b1197266d1a50ecc9187a91e481/Modules/_io/bufferedio.c

So, dooing a little more tracing
```py
import io
import tempfile
print(io.DEFAULT_BUFFER_SIZE)

with tempfile.NamedTemporaryFile() as file:
    pass
```

`strace --trace=file --no-abbrev  python temp_test.py`

shows 
```
openat(AT_FDCWD, "/run/user/1000/tmp486dtl2p", O_RDWR|O_CREAT|O_EXCL|O_NOFOLLOW|O_CLOEXEC, 0600) = 3
newfstatat(3, "", {st_dev=makedev(0, 0x3b), st_ino=1329, st_mode=S_IFREG|0600, st_nlink=1, st_uid=1000, st_gid=100, st_blksize=4096, st_blocks=0, st_size=0, st_atime=1673231853 /* 2023-01-08T20:37:33.136371805-0600 */, st_atime_nsec=136371805, st_mtime=1673231853 /* 2023-01-08T20:37:33.136371805-0600 */, st_mtime_nsec=136371805, st_ctime=1673231853 /* 2023-01-08T20:37:33.136371805-0600 */, st_ctime_nsec=136371805}, AT_EMPTY_PATH) = 0
unlink("/run/user/1000/tmp486dtl2p")    = 0
```

On my btrfs, it got `st_blksize=4096`, so even less than the default.

I assume in whatever environment you're seeing failures, it must be considerably larger (at least > 90350 for python to not have written through anything at all), since BufferedWriter should flush through:

https://docs.python.org/3/library/io.html#io.BufferedWriter
> - when the buffer gets too small for all pending data
> -  when [flush()](https://docs.python.org/3/library/io.html#io.BufferedWriter.flush) is called;
> - when a seek() is requested (for [BufferedRandom](https://docs.python.org/3/library/io.html#io.BufferedRandom) objects);
> - when the [BufferedWriter](https://docs.python.org/3/library/io.html#io.BufferedWriter) object is closed or destroyed.

Comment (puetzk) [reply pkgs/tools/backup/borgbackup/default.nix:38]: /run is a tmpfs, so, by process of elimination - you are probably using a kernel with different transparent hugepage settings: https://www.kernel.org/doc/html/v5.9/admin-guide/mm/transhuge.html#hugepages-in-tmpfs-shmem, as tmpfs uses the memory page size for st_blksize. Which would be one of the few things nix hasn't locked down for reproducibility...

Comment (mweinelt) [reply pkgs/tools/backup/borgbackup/default.nix:38]: I can reproduce the issue on one of my remote builders, that does not use `tmpfs` for `/tmp`, because nix dumps pretty large things into there sometimes and doesn't always clean up after itself.

On my desktop however I have a tmpfs and the test works there.

**Comment by mweinelt** [comment]
Dropped the patches. Moved library discovery to pkgconfig. Some love for libb2, which isn't actually used anymore.

**Comment by mweinelt** [comment]
The aarch64-linux community builder shows the same problem, so we should probably apply the patch after all.

File: pkgs/tools/backup/borgbackup/default.nix (hunk 1 above)
Comment (puetzk) [reply pkgs/tools/backup/borgbackup/default.nix]: tmpfs without hugetlb (the default) has 4kiB pages, as I'm seeing and that "works" (though there should indeed be a flush() in the code, that's a legit bug in the test now). I was just continuing to dig at the non-determinism, since that's also unexpected in an environment as tightly-controlled as nix, and I didn't want to be the one trying to open an upsteam ticket for a problem I didn't understand and couldn't reproduce :-)

It seems like environment you're seeing failures on (apparently zfs) must have `st_blksize` > 90350 for python not to have written anything at all. https://docs.oracle.com/cd/E23824_01/html/821-1448/gazss.html says "The default block size for volumes is 8 KB. Any power of 2 from 512 bytes to 128 KB is valid." but it seems to be something set at volume creation time.

I guess you could check with strace and we'd know for sure, but finding out cpython's behavior is based on st_blksize is probably enough to know where something the derivation doesn't control could come from.

**Comment by puetzk** [comment]
I think the correct fix would be:

```diff
--- a/src/borg/testsuite/hashindex.py
+++ b/src/borg/testsuite/hashindex.py
@@ -199,6 +199,7 @@ def test_size_on_disk_accurate(self):
         with tempfile.NamedTemporaryFile() as file:
             idx.write(file)
+            file.flush()
             size = os.path.getsize(file.fileno())
         assert idx.size() == size
 
```

File: pkgs/tools/backup/borgbackup/default.nix (hunk 1 above)
Comment (mweinelt) [reply pkgs/tools/backup/borgbackup/default.nix]: Opened https://github.com/borgbackup/borg/issues/7250 and speculated it is related to ZFS' 5 second commit interval, but since you dug deeper you might as well chime in.

**Comment by mweinelt** [comment]
Feel free to pull request it upstream and we can fetch that instead.

**Comment by puetzk** [comment]
yeah, I was just hoping to figure out a setup that actually fails, so that I could test that this patch indeed fixes it, before sending a PR. It *should*, but it wasn't failing for me in the first place :-)

**Comment by puetzk** [comment]
If you have currently have a setup that does, it would be nice to try the patch (or at least do the `strace --trace=file --no-abbrev  python temp_test.py` and confirm it's showing a large `st_blksize` so we know the theory is right)

**Comment by mweinelt** [comment]
Tested on both machines, where it failed before.

**Comment by puetzk** [comment]
https://github.com/borgbackup/borg/pull/7252 has the 1.2-maint commit:
```nix
  patches = [
    # Fix HashIndexSizeTestCase.test_size_on_disk_accurate problems on ZFS, borgbackup/borg#7250
    (fetchpatch {
      url = "https://github.com/borgbackup/borg/commit/537a814e53e20013a041faa7192da005f137cf5b.patch";
      hash = "sha256-dnF/FW8pS4Ub9aAL4b7zf6ZNjMZaiMqdtl5R+DlAZTM=";
    })
  ];
```
